src/shogun/kernel/GaussianKernel.cpp

```diff
@@ -88,21 +91,42 @@ void CGaussianKernel::set_width(float64_t w)
 	m_log_width = std::log(w / 2.0) / 2.0;
 }
 
+auto CGaussianKernel::kernel_function(int32_t idx_a, int32_t idx_b) const
+{
+	// this could be written as Eigen::Matrix<float64_t, n_differentiable_params, 1>;
+	using EigenScalar = Eigen::Matrix<float64_t, 1, 1>;
+	Eigen::AutoDiffScalar<EigenScalar> eigen_log_width(m_log_width);
+	// resize(n_differentiable_params)
+	eigen_log_width.derivatives().resize(1);
+	// this could be written as 
+	// eigen_log_width.derivatives() = EigenScalar::Unit(1,i);
+	// where i is the idx of the adjoint
+	eigen_log_width.derivatives() = EigenScalar::Unit(1,0);
+	auto el = CShiftInvariantKernel::distance(idx_a, idx_b);
+	return exp(-el / (exp(eigen_log_width * 2.0) * 2.0));
+}
+
 SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(const TParameter* param, index_t index)
 {
+	using std::exp;
+
 	require(lhs, "Left hand side features must be set!");
-	require(rhs, "Rightt hand side features must be set!");
+	require(rhs, "Right hand side features must be set!");
 
 	if (!strcmp(param->m_name, "log_width"))
```

Comment:
so I guess a next step could be to start thinking about getting rid of this explicit code, and rather automatically offer this derivative through registering something in the ctors ....

Reply:
I think for that we should maybe have a base class for classes that have parameters that we can take the derivative wrt. This class registers the gradient parameters in some vector and then we can get the index from there.
Basically when we do watch_param(...) this would add the variable in such a vector if it has the flag GRADIENT